### PR TITLE
WT-7619 Add a new optimization to skip pages in cursor traversal when all entries on the page are deleted

### PR DESCRIPTION
**Pull request by hbokhari**

### CODE REVIEW

File: src/btree/bt_curnext.c [hunk 1]
```diff
@@ -677,6 +679,23 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
...
         page = cbt->ref == NULL ? NULL : cbt->ref->page;
 
+        /*
+         * Determine if all records on the page have been deleted and all the tombstones are visible
+         * to our transaction. If so, we can avoid reading the records on the page and move to the
+         * next page. We base this decision on the aggregate stop point added to the page during the
+         * last reconciliation. We can skip this test if the page has been modified since it was
+         * reconciled or the underlying cursor is configured to ignore tombstones.
+         */
+        if (session->txn->isolation == WT_ISO_SNAPSHOT &&
```
Comment (quchenhao) [inline src/btree/bt_curnext.c:689]: I think we need to exclude this optimization for history store? The history store dump with globally visible tombstone will be broken by this change I think.

File: src/btree/bt_curprev.c [hunk 2]
```diff
@@ -624,6 +626,23 @@ __wt_btcur_prev_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
...
+         * to our transaction. If so, we can avoid reading the records on the page and move to the
+         * next page. We base this decision on the aggregate stop point added to the page during the
+         * last reconciliation. We can skip this test if the page has been modified since it was
+         * reconciled or the underlying cursor is configured to ignore tombstones.
+         */
+        if (session->txn->isolation == WT_ISO_SNAPSHOT &&
+          !F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE) && cbt->ref != NULL &&
+          __wt_ref_addr_copy(session, cbt->ref, &addr) == true && page != NULL &&
+          !__wt_page_is_modified(page) &&
+          __wt_txn_visible(session, addr.ta.newest_stop_txn, addr.ta.newest_stop_ts) &&
```
Comment (quchenhao) [inline src/btree/bt_curprev.c:640]: Can we create a inline function for visibility checking of the aggregated time window similar to time window, like __wt_txn_tw_stop_visible or do we already have one?

Comment (hbokhari) [reply src/btree/bt_curprev.c]: I believe there is no existing inline function that checks time aggregate data. The `__wt_txn_visible()` is already an inline function so I am assuming we will not gain much with making another enclosing inline function for time aggregate.

Comment (quchenhao) [reply src/btree/bt_curprev.c]: If there is one, why we are not using it? We should standardize the way we do visibility checking.

Comment (hbokhari) [reply src/btree/bt_curprev.c]: Sorry if I wasn't clear - there is no existing inline function that checks time aggregate window.

File: src/include/btree_inline.h [hunk 3]
```diff
@@ -2018,3 +2018,56 @@ __wt_bt_col_var_cursor_walk_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *
...
+     * our transaction. If so, we can avoid reading the records on the page and move to the next
+     * page. We base this decision on the aggregate stop point added to the page during the last
+     * reconciliation. We can skip this test if the page has been modified since it was reconciled
+     * or the underlying cursor is configured to ignore tombstones.
+     *
+     * We are making these decisions while holding a lock for the page as checkpoint or eviction can
+     * make changes to the data structures (i.e., aggregate timestamps) we are reading.
+     */
+    if (session->txn->isolation == WT_ISO_SNAPSHOT &&
+      !F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE) && previous_state == WT_REF_MEM) {
```
Comment (hbokhari) [inline src/include/btree_inline.h:2058]: @quchenhao History store cursors set the flag `WT_CURSTD_IGNORE_TOMBSTONE` on the underlying cursor.

File: src/btree/bt_curprev.c (hunk 2 above)
Comment (quchenhao) [reply src/btree/bt_curprev.c]: got it

File: src/btree/bt_curnext.c [hunk 4]
```diff
@@ -730,7 +741,7 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
                 continue;
             }
         }
-
+skip_page:
```
Comment (sulabhM) [inline src/btree/bt_curnext.c:744]: Should "skip_page" goto label jump to line 766 instead, i.e. to:
` WT_ERR(__wt_tree_walk(session, &cbt->ref, flags));`

Why check for deleted records on the page we skipped reading, they will always be 0. Am I missing something ?

File: src/btree/bt_curprev.c [hunk 5]
```diff
@@ -678,7 +689,7 @@ __wt_btcur_prev_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
             if (ret != WT_NOTFOUND)
                 break;
         }
-
+skip_page:
```
Comment (sulabhM) [inline src/btree/bt_curprev.c:692]: As in the case of next, we should skip checking for delete count if we did not even read the page.

File: src/include/btree_inline.h [hunk 6]
```diff
@@ -2018,3 +2018,56 @@ __wt_bt_col_var_cursor_walk_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *
...
+    WT_REF *ref;
+    WT_SESSION_IMPL *session;
+    uint8_t previous_state;
+    bool can_skip;
+
+    session = CUR2S(cbt);
+    ref = cbt->ref;
+    page = cbt->ref == NULL ? NULL : cbt->ref->page;
+
+    if (ref == NULL || page == NULL)
```
Comment (sulabhM) [inline src/include/btree_inline.h:2041]: Line 2039, the assignment for page takes care of ref being NULL. If ref is NULL, page will be NULL, so you can just check for page to be NULL here.

File: src/include/btree_inline.h [hunk 7]
```diff
@@ -2018,3 +2018,56 @@ __wt_bt_col_var_cursor_walk_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *
...
+     * make changes to the data structures (i.e., aggregate timestamps) we are reading.
+     */
+    if (session->txn->isolation == WT_ISO_SNAPSHOT &&
+      !F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE) && previous_state == WT_REF_MEM) {
+
+        /* We only try to lock the page once. */
+        if (!WT_REF_CAS_STATE(session, ref, previous_state, WT_REF_LOCKED))
+            return false;
+
+        if (!__wt_page_is_modified(page) && __wt_ref_addr_copy(session, ref, &addr) == true &&
```
Comment (sulabhM) [inline src/include/btree_inline.h:2064]: What happens if the page is modified. The modification should be newer than the aggregated information. If based on the aggregated information our reader can skip the page, a modification that comes after should not make a difference. Am I missing something here ?

Comment (sulabhM) [inline src/include/btree_inline.h:2064]: Since __wt_ref_addr_copy returns a boolean, the convention is to use the result directly, i.e.

`.. && __wt_ref_addr_copy(session, ref, &addr) && ..`

File: test/suite/test_cursor17.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+        # Insert few thousand key-value pairs.
+        for key in range(total_keys):
+            self.session.begin_transaction()
+            cursor[key] = value1
+            self.session.commit_transaction('commit_timestamp=' + timestamp_str(commit_timestamp))
+            commit_timestamp += 1
+
+        # Delete everything on the table except for the first and the last KV pair.
+        for key in range(total_keys):
```
Comment (sulabhM) [inline test/suite/test_cursor17.py:73]: can we adjust the for loop range such that we do not need the if condition, so if we change to:
`for key in range(1, total_keys - 1):`

We will not need the if condition.

**Review by sulabhM — CHANGES_REQUESTED** [review]

File: src/btree/bt_curnext.c (hunk 4 above)
Comment (hbokhari) [reply src/btree/bt_curnext.c]: Fixed.

File: src/btree/bt_curprev.c (hunk 5 above)
Comment (hbokhari) [reply src/btree/bt_curprev.c]: Fixed.

File: src/include/btree_inline.h (hunk 6 above)
Comment (hbokhari) [reply src/include/btree_inline.h]: Fixed.

File: src/include/btree_inline.h (hunk 7 above)
Comment (hbokhari) [reply src/include/btree_inline.h]: Fixed.

File: test/suite/test_cursor17.py (hunk 8 above)
Comment (hbokhari) [reply test/suite/test_cursor17.py]: Good catch. I have change the outer `for` loop.

File: src/include/btree_inline.h (hunk 7 above)
Comment (hbokhari) [reply src/include/btree_inline.h]: There is always a chance that a transaction has modified the page between setting aggregate and this check. Imagine a case where aggregate contains stop txn id 10. Then a new transaction with id 11 inserted a key on the page. Now if the current session's txn id is greater than 11 and transaction id 11 committed before the current transaction started, the value inserted by txn id 11 should be visible.

Comment (sulabhM) [reply src/include/btree_inline.h]: That sounds good to me. I am convinced we need the is-modified check.

File: dist/stat_data.py [hunk 9]
```diff
@@ -812,11 +812,13 @@ def __init__(self, name, desc, flags=''):
     CursorStat('cursor_next_hs_tombstone', 'cursor next calls that skip due to a globally visible history store tombstone'),
     CursorStat('cursor_next_skip_ge_100', 'cursor next calls that skip greater than or equal to 100 entries'),
     CursorStat('cursor_next_skip_lt_100', 'cursor next calls that skip less than 100 entries'),
+    CursorStat('cursor_next_skip_pages', 'cursor next call skipped pages without reading'),
```
Comment (sulabhM) [inline dist/stat_data.py:815]: Maybe we can be more clear on the description. It is a count of pages that next call(s) are skipping without reading. It is not a count of next(s). Maybe reword as:

"Total number of pages skipped without reading by cursor next calls"

Similarly for the prev calls.

**Review by sulabhM — APPROVED** [review]
Other than a comment about rewording the statistics, this change looks good to me.

**Comment by sulabhM** [comment]
@hbokhari , Once all LGTMs are obtained, I recommend we should run a full WT test suite, stress tests, MongoDB patch test and all MongoDB perf tests before we checkin.

**Comment by hbokhari** [comment]
@keithbostic, this is the comment on `__wt_page_is_modified()`
    /*
     * Be cautious modifying this function: it's reading fields set by checkpoint reconciliation,
     * and we're not blocking checkpoints (although we must block eviction as it might clear and
     * free these structures).
     */
Is there a guarantee that page modify structures would remain intact while we are reading them outside locks?

**Comment by keithbostic** [comment]
> @keithbostic, this is the comment on `__wt_page_is_modified()`
> /*
> * Be cautious modifying this function: it's reading fields set by checkpoint reconciliation,
> * and we're not blocking checkpoints (although we must block eviction as it might clear and
> * free these structures).
> */
> Is there a guarantee that page modify structure would remain intact while we are reading it outside locks?

@hbokhari, the only thing that can discard the modify structure is eviction, so as long as we're holding a hazard reference we should be fine.

File: dist/stat_data.py (hunk 9 above)
Comment (tammybailey) [reply dist/stat_data.py:815]: I'd suggest renaming the statistic to `cursor_next_skip_page_count` or `cursor_next_skip_page_total` for clarity, and agree with Sulabh's suggestion to update the description.

**Comment by hbokhari** [comment]
> > @keithbostic, this is the comment on `__wt_page_is_modified()`
> > /*
> > 
> > * Be cautious modifying this function: it's reading fields set by checkpoint reconciliation,
> > * and we're not blocking checkpoints (although we must block eviction as it might clear and
> > * free these structures).
> >   */
> >   Is there a guarantee that page modify structure would remain intact while we are reading it outside locks?
> 
> @hbokhari, the only thing that can discard the modify structure is eviction, so as long as we're holding a hazard reference we should be fine.

Thanks. I have fixed this in 8bf72a3.

File: dist/stat_data.py (hunk 9 above)
Comment (hbokhari) [reply dist/stat_data.py]: Fixed in a285b69.

File: test/suite/test_cursor17.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+import wiredtiger, wttest
+from wtscenario import make_scenarios
+from wiredtiger import stat, WT_NOTFOUND
+
+def timestamp_str(t):
+    return '%x' % t
+
+# test_cursor17.py
+# Test optimization that skips page during cursor traversal if it determines that all records on the
```
Comment (tammybailey) [inline test/suite/test_cursor17.py:37]: It would be useful to add more details in the test description. We are specifically targeting the mass deletion scenario, and we expect that a "significant" number of pages are skipped in this scenario. Can we better quantify "significant", perhaps as a percentage of the `total_keys` or total number of pages, rather than choosing `1000`?

Suggesting the following:
_"Test the cursor traversal optimization for delete heavy workloads where that a significant number of pages are skipped when all records on the page are deleted with a tombstone visible to the current transaction."_

And add in something along the lines of:
_"We expect that X% pages are skipped when deleting Y% of the keys."_

File: test/suite/test_cursor17.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        total_keys = 40000
+
+        # Keep oldest timestamp pinned.
+        self.conn.set_timestamp('oldest_timestamp=' + timestamp_str(1))
+        self.conn.set_timestamp('stable_timestamp=' + timestamp_str(2))
+        cursor = self.session.open_cursor(uri)
+
+        commit_timestamp = 2
+
+        # Insert few thousand key-value pairs.
```
Comment (tammybailey) [inline test/suite/test_cursor17.py:65]: This comment is misleading as we are inserting 40 thousand KVs which is more than "a few".

File: test/suite/test_cursor17.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        # Position the cursor on the first record.
+        cursor.set_key(0)
+        self.assertEqual(cursor.search(), 0)
+        # This should move the cursor to the last record.
+        self.assertEqual(cursor.next(), 0)
+        self.assertEqual(cursor.get_key(), total_keys-1)
+
+        # Check if we skipped any pages while moving the cursor.
+        #
+        # I ran this test few times and the stat appeared to be in range of ~1400. I have
```
Comment (tammybailey) [inline test/suite/test_cursor17.py:93]: See my first comment - is there a better way to choose the comparison number? Also, please move the explanation for the number chosen to the beginning of the test.

File: test/suite/test_cursor17.py (hunk 11 above)
Comment (hbokhari) [reply test/suite/test_cursor17.py:69]: Fixed.

File: test/suite/test_cursor17.py (hunk 10 above)
Comment (hbokhari) [reply test/suite/test_cursor17.py]: I have updated comments in ef0d1ba.

File: test/suite/test_cursor17.py (hunk 12 above)
Comment (hbokhari) [reply test/suite/test_cursor17.py]: I have changed the method for calculating the expected number of skipped pages. Have a look at d5b6ef6.

**Comment by hbokhari** [comment]
@quchenhao @tammybailey @agorrod, can you have a look?

File: test/suite/test_cursor17.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        # Check if we skipped any pages while moving the cursor.
+        #
+        # We calculate the number of pages we expect to skip based on the total number of leaf pages
+        # reported in the WT stats. We subtract 2 from the count of leaf pages in the table and test
+        # that we atleast skipped 80% of the expected number of pages.
+
+        leaf_pages_in_table = self.get_stat(stat.dsrc.btree_row_leaf, uri)
+        expected_pages_skipped = ((leaf_pages_in_table - 2) * 8) // 10
+        skipped_pages = self.get_stat(stat.conn.cursor_next_skip_page_count, None)
+        self.assertGreater(skipped_pages, expected_pages_skipped)
```
Comment (quchenhao) [inline test/suite/test_cursor17.py:104]: Is this test reliable? I think if we fail to lock the page, the number of skipped pages would be less than expected, is that correct?

Comment (hbokhari) [reply test/suite/test_cursor17.py:104]: The cache size used for this test is 50MB and table size is around 24MB. Therefore, I expect minimal eviction activity once I have performed a checkpoint. I have kept a 20% leeway in case there are pages that we somehow are unable to skip.

Comment (quchenhao) [reply test/suite/test_cursor17.py:104]: I think we only need to assert the page skip happens and that should be safer? I don't understand why we need to assert that we skip a certain amount of pages to test the feature is working. I think this is not a performance test, isn't it?

Comment (tammybailey) [reply test/suite/test_cursor17.py:104]: The intent is to test the optimization. Arguably this may not be the best or most reliable approach. I believe there is follow-on work to add this scenario to our `workgen` performance testing - @hbokhari do you feel that would be a more appropriate place to add this testing?

Comment (hbokhari) [reply test/suite/test_cursor17.py:104]: I believe performance testing with `workgen` would indicate if this optimization is giving us any benefit with certain workloads. Here, I am functionally testing that optimization is working as intended ,i.e., it is able to skip pages given certain conditions such as larger than table cache and a clean tree in cache (no eviction activity). If the assumption here that optimization should skip all pages between first and the last page becomes invalid, we might want to go back and review the optimization.

Having said that, if a majority of reviewer feels the test is too strict, I am happy to remove the constraint.

Comment (tammybailey) [reply test/suite/test_cursor17.py:104]: @hbokhari I'm happy to trust your judgement on this as you've been working on the issue for several months.

**Review by tammybailey — APPROVED** [review]
lgtm

**Review by quchenhao — APPROVED** [review]
lgtm
